dbms/src/DataStreams/AggregatingSortedBlockInputStream.h

```diff
@@ -63,6 +64,9 @@ class AggregatingSortedBlockInputStream : public MergingSortedBlockInputStream
      */
     void insertSimpleAggregationResult(MutableColumns & merged_columns);
 
+    /// Memory pool for SimpleAggregateFunction.
+    Arena arena;
```

Comment:
Although I'm not sure about this bit

Reply:
It will accumulate data without clearing, that will result in memory leak.
We should clear arena after processing every aggregate function state.

Reply:
My code was "inspired" by SummingSortedBlockInputStream.cpp :

```
            // Specialized case for unary functions
            if (desc.column_numbers.size() == 1)
            {
                auto & col = cursor->all_columns[desc.column_numbers[0]];
                desc.add_function(desc.function.get(), desc.state.data(), &col, cursor->pos, nullptr);
            }
            else
            {
                // Gather all source columns into a vector
                ColumnRawPtrs columns(desc.column_numbers.size());
                for (size_t i = 0; i < desc.column_numbers.size(); ++i)
                    columns[i] = cursor->all_columns[desc.column_numbers[i]];

                desc.add_function(desc.function.get(), desc.state.data(), columns.data(), cursor->pos, nullptr);
            }
```

Should this be fixed too?

Reply:
Looks like it must be fixed too. (Also look at AggregatingMergeTree just in case).
But first let's try to provide a failing test.

Reply:
>It will accumulate data without clearing, that will result in memory leak.
We should clear arena after processing every aggregate function state.

I was thinking that Arena will take care of this, seems that I was wrong, okay I will take a closer look and take care of two new codepath

>But first let's try to provide a failing test.

There is a test for the SimpleAggregateFunction, and it is pretty simple, so that said that if AggregatingMergeTree has these problems I guess that this will be triggered by someone already.

Reply:
> It will accumulate data without clearing, that will result in memory leak.
We should clear arena after processing every aggregate function state.

Fixed in the dc6431d, can you take a look?

> (Also look at AggregatingMergeTree just in case)

Looks like it is safe (at least there is no `nullptr` passing)

> SummingSortedBlockInputStream

Looks like it is safe too, since `nullptr` passed only for `sumWithOverflow` which do not use `arena` (but maybe I missing something?)

Reply:
Another small change...

There is `IAggregateFunction::allocatesMemoryInArena`. We should check if any of the aggregate functions requires Arena (store it in a bool flag) and create Arena only in that case.

Reply:
> Looks like it is safe too, since nullptr passed only for sumWithOverflow which do not use arena (but maybe I missing something?)

That's Ok.

Reply:
> There is IAggregateFunction::allocatesMemoryInArena. We should check if any of the aggregate functions require Arena (store it in a bool flag) and create Arena only in that case.

- 4252ffae8e Allocates arena for SimpleAggregateFunction only if IAggregateFunction requires
- 18621c1a48 Set allocatesMemoryInArena for SingleValueDataString to true
 
*By the way, before merging, I would like to squash first two patches*

Reply:
Ok. Thank you! Happy that another issue in our codebase was fixed.

> I would like to squash first two patches

Do you want us to squash all commits before merge or you'll squash some commits in your branch? PS. We take it easy: no problem if some commits in master will contain wrong code.

Reply:
> Do you want us to squash all commits before merge or you'll squash some commits in your branch?

Not all, just the first two of them, rebased

> PS. We take it easy: no problem if some commits in master will contain wrong code.

Yeah, but I like to keep the history clean.

Reply:
And thank you for your comments!